// staticmaps/dist/bound.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _bbox = _interopRequireDefault(require("@turf/bbox"));

var Circle = /*#__PURE__*/function () {
  function Circle() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck2["default"])(this, Circle);
    this.options = options;
    this.coords = this.options.coords;
  }
  /**
   * calculate the coordinates of the envelope / bounding box: (min_lon, min_lat, max_lon, max_lat)
   */


  (0, _createClass2["default"])(Circle, [{
    key: "extent",
    value: function extent() {
      var line = {
        type: 'LineString',
        coordinates: this.coords
      };
      return (0, _bbox["default"])(line);
    }
  }]);
  return Circle;
}();

exports["default"] = Circle;
